plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php

```diff
@@ -186,42 +186,61 @@ public function edit_account_form_fields() {
 	 * @param integer $user_id User ID.
 	 */
 	public function save_account_form_fields( $user_id ) {
-		// phpcs:disable WordPress.Security.NonceVerification.Missing
-		$customer          = new WC_Customer( $user_id );
-		$additional_fields = $this->checkout_fields_controller->get_fields_for_location( 'contact' );
-		$field_values      = array();
 
-		foreach ( array_keys( $additional_fields ) as $key ) {
-			$post_key = CheckoutFields::get_group_key( 'other' ) . $key;
-			if ( ! isset( $_POST[ $post_key ] ) ) {
-				continue;
-			}
+		try {
+			$customer = new WC_Customer( $user_id );
 
-			$field_value = $this->checkout_fields_controller->sanitize_field( $key, wc_clean( wp_unslash( $_POST[ $post_key ] ) ) );
-			$validation  = $this->checkout_fields_controller->validate_field( $key, $field_value );
+			$additional_fields = $this->checkout_fields_controller->get_fields_for_location( 'contact' );
 
-			if ( is_wp_error( $validation ) && $validation->has_errors() ) {
-				wc_add_notice( $validation->get_error_message(), 'error' );
-				continue;
+			$field_values = array();
+
+			foreach ( $additional_fields as $key => $field_data ) {
+				// We can't skip, field might be required.
+				$field_value = trim( wc_clean( wp_unslash( $_POST[ $key ] ?? '' ) ) ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
+
+				if ( 0 === strlen( $field_value ) && ! ( $field_data['required'] ?? false ) ) {
+					$field_values[ $key ] = '';
+					continue; // Skip, otherwise validation might prevent saving empty optional fields.
+				}
+
+				$validation = $this->checkout_fields_controller->validate_field( $key, $field_value );
+				// Sanitize should always be after validation to ensure char stripping doesn't affect validation.
+				$field_value = $this->checkout_fields_controller->sanitize_field( $key, $field_value );
```

Comment:
Why bother with sanitization before checking if validation worked? I suspect this approach is incorrect also. Validation runs after sanitize in API.

Reply:
I think validation should run before sanitization, as validation will expose errors whereas sanitizing just ensures data is safe to use/store. There seems to be some  [cases] around (https://github.com/woocommerce/woocommerce/blob/7d74306c34520d5683d4a76e761146c5557cff75/plugins/woocommerce/includes/class-wc-checkout.php#L898) and the developer.wordpress.org [states it should be done as early as possible](https://developer.wordpress.org/apis/security/data-validation/), and escape as late as possible, not being too specific where would sanitize land 🤷
If we're running validation to check allowed chars after sanitizing those chars, will it not always pass? I had issues with this recently after sanitizing first 😅